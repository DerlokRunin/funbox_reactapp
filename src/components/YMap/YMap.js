import React, { Component } from 'react';
import PropTypes from "prop-types";
import { YMaps, Map, Placemark, Polyline } from 'react-yandex-maps';

class YMap extends Component {

    static propTypes = {
        mapState: PropTypes.shape({ //начальные координаты карты
            center: PropTypes.array.isRequired,
            zoom: PropTypes.number.isRequired
        }),
        points: PropTypes.array,//точки с координатами
        onGeometryChange: PropTypes.func, //действия при изменение положения точки
        onDragMap: PropTypes.func//действия при смещении центра карты
    };

    constructor(props){
        super(props);

        this.myMap= React.createRef();
        this.addEventDrag = false;
    }

    render() {
        const {mapState, points, onGeometryChange} = this.props;

        //записываем все координаты
        this.coordinates = [];
        for (let i = 0; i < points.length; i++){
            this.coordinates.push(points[i].coordinates);
        }

        //Перебираем маркеры
        const markerElements = points.map(point =>
            <Placemark key={point.id}
                geometry={{
                    type: 'Point',
                    coordinates: point.coordinates,
                }}
                properties={{
                    iconContent: point.title,
                    hintContent: point.title,
                    balloonContent: point.title,
                }}
                options={{
                    preset: 'islands#blackStretchyIcon',
                    draggable: true,
                }}
                onGeometryChange={onGeometryChange.bind(this, point.id)}
            />
        );
        return(
            <YMaps>
                <Map state={mapState} width="100%" height="100%" ref={this.myMap} onMouseDown={this.onMapClick}>
                    {markerElements}
                    <Polyline
                        geometry={{
                            type: 'LineString',
                            coordinates: this.coordinates,
                        }}
                        properties={{
                            hintContent: 'линия',
                        }}
                    />
                </Map>
            </YMaps>
        );
    }



    //кликнули по карте
    onMapClick = () => {
        if(!this.addEventDrag){
            const {onDragMap} = this.props;
            try {
                this.addEventDrag = true;
                //событие перемещения по карте
                let map = this.myMap.current.state.instance;
                map.events.add('actionend', function (e) {
                    //смещаем центр карты
                    var pixelCenter = map.getGlobalPixelCenter();
                    var geoCenter = map.options.get('projection').fromGlobalPixels(pixelCenter, map.getZoom());
                    //смещаем центр карты
                    onDragMap(geoCenter);
                });
            }catch(e){}
        }
        console.log();
    }
}

export default YMap;