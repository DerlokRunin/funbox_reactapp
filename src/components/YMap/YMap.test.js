import React from 'react';
import YMap from './YMap';


it('проверка рендеринга yandex map', () => {
    let points = [
        {id:1, title:'test', coordinates: [42.423, 35.353] },
        {id:2, title:'test2', coordinates: [42.423, 35.353] },
        {id:3, title:'test3', coordinates: [42.423, 35.353] }
    ];
    let mapState = {center: [42.423, 35.353], zoom: 14};

    const wrapper = shallow(
        <YMap mapState={mapState}
              points={points}
              onGeometryChange={function () {}}
              onDragMap={function () {}}
        />
    );
    expect(wrapper).toMatchSnapshot();
});