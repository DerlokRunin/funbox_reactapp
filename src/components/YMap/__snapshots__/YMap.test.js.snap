// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`проверка рендеринга yandex map 1`] = `
<YMaps
  enterprise={false}
  onApiAvaliable={[Function]}
  version="2.1"
>
  <Map
    height="100%"
    instanceRef={[Function]}
    onMouseDown={[Function]}
    state={
      Object {
        "center": Array [
          42.423,
          35.353,
        ],
        "zoom": 14,
      }
    }
    width="100%"
  >
    <Placemark
      geometry={
        Object {
          "coordinates": Array [
            42.423,
            35.353,
          ],
          "type": "Point",
        }
      }
      key="1"
      onGeometryChange={[Function]}
      options={
        Object {
          "draggable": true,
          "preset": "islands#blackStretchyIcon",
        }
      }
      properties={
        Object {
          "balloonContent": "test",
          "hintContent": "test",
          "iconContent": "test",
        }
      }
    />
    <Placemark
      geometry={
        Object {
          "coordinates": Array [
            42.423,
            35.353,
          ],
          "type": "Point",
        }
      }
      key="2"
      onGeometryChange={[Function]}
      options={
        Object {
          "draggable": true,
          "preset": "islands#blackStretchyIcon",
        }
      }
      properties={
        Object {
          "balloonContent": "test2",
          "hintContent": "test2",
          "iconContent": "test2",
        }
      }
    />
    <Placemark
      geometry={
        Object {
          "coordinates": Array [
            42.423,
            35.353,
          ],
          "type": "Point",
        }
      }
      key="3"
      onGeometryChange={[Function]}
      options={
        Object {
          "draggable": true,
          "preset": "islands#blackStretchyIcon",
        }
      }
      properties={
        Object {
          "balloonContent": "test3",
          "hintContent": "test3",
          "iconContent": "test3",
        }
      }
    />
    <Polyline
      geometry={
        Object {
          "coordinates": Array [
            Array [
              42.423,
              35.353,
            ],
            Array [
              42.423,
              35.353,
            ],
            Array [
              42.423,
              35.353,
            ],
          ],
          "type": "LineString",
        }
      }
      properties={
        Object {
          "hintContent": "линия",
        }
      }
    />
  </Map>
</YMaps>
`;
